File: go_echo/model/loan/loantran.go
package loan

import (
	"math"
	"time"

	"../../infrastructure/loan"
)

type LoanTran struct {
	ID        int64  `json:"id"`
	Mouth     string `json:"mouth"`
	Total     int64  `json:"total"`
	Principal int64  `json:"principal"`
	Interest  int64  `json:"interest"`
	Balance   int64  `json:"balance"`
}

// 元利均等ローン計算
func CalcLoan(borrowingAmount int64, repaymentPeriod int64) []LoanTran {
	r := &loan.RepositoryImp{}
	interestRate, _ := r.FindRate()
	fixedRepaymentAmount := calcFixedRepaymentAmount(borrowingAmount, interestRate, repaymentPeriod)

	repaymentMouthCount := repaymentMouthCount(repaymentPeriod)
	balanceMinus := balancesMinus(borrowingAmount)
	balance := borrowingAmount
	date := time.Now().UTC()
	loanTrans := []LoanTran{}

	for i := 1; i < int(repaymentMouthCount + 1); i++ {
		loanTran := buildLoanTran(int64(i), date.Format("2006/01"), fixedRepaymentAmount, interestRate, balance)
		loanTrans = append(loanTrans, loanTran)
		date = date.AddDate(0, 1, 0)
		balance = balanceMinus(loanTran.Principal)
	}
	return loanTrans
}

// 本来ここに記載するものではないが、簡略化のため
func RoundDown(num float64, place int64) float64 {
	shift := math.Pow(10, float64(place))
	return math.Floor(num*shift) / shift
}

// 年利から月ごとの金利を計算
func mouthRate(interestRate float64) float64 {
	return RoundDown(interestRate/100/12, 10)
}

// 年数から支払回数を計算
func repaymentMouthCount(repaymentPeriod int64) int64 {
	return repaymentPeriod * 12
}

// 現在の残高と年利から今月の支払い利息を計算
func calcMonthInterest(balance int64, interestRate float64) int64 {
	return int64(RoundDown(float64(balance)*mouthRate(interestRate), 0))
}

// 毎月の一定返済金額を計算する
func calcFixedRepaymentAmount(balance int64, interestRate float64, repaymentPeriod int64) int64 {
	repaymentMouthCount := repaymentMouthCount(repaymentPeriod)
	mouthRate := mouthRate(interestRate)

	numerator := RoundDown(float64(balance)*mouthRate*(math.Pow((float64(1)+mouthRate), float64(repaymentMouthCount))), 10)
	denominator := RoundDown(math.Pow((float64(1)+mouthRate), float64(repaymentMouthCount))-float64(1), 10)
	return int64(RoundDown(numerator/denominator, 0))
}

func buildLoanTran(id int64, month string, total int64, interestRate float64, beforeBalance int64) LoanTran {
	interest := calcMonthInterest(beforeBalance, interestRate)
	principal := total - interest
	// 支払後残高が今回支払額より少ない場合は最終支払分と判断しまとめる
	lastBalance := int64(0)
	if beforeBalance-principal-principal < 0 {
		lastBalance = beforeBalance - principal
	}

	return LoanTran{
		ID:        id,
		Mouth:     month,
		Total:     total + lastBalance,
		Principal: principal + lastBalance,
		Interest:  interest,
		Balance:   beforeBalance - principal - lastBalance,
	}
}

File: ts_express/src/model/loan/LoanTran.ts
import { Rate } from '../../infrastructure/loan/Rate';
import { balancesMinus } from './Balance';

export class LoanTran {
    private id: number;
    private mouth: string;
    private total: number;
    private principal: number;
    private interest: number;
    private balance: number;

    constructor(id: number,
        mouth: string,
        total: number,
        principal: number,
        interest: number,
        balance: number) {

        this.id = id;
        this.mouth = mouth;
        this.total = total;
        this.principal = principal;
        this.interest = interest;
        this.balance = balance;
    }
}

export const calcLoan = (borrowingAmount: number, repaymentPeriod: number): LoanTran[] => {
    let rateInterface: RateInterface;
    rateInterface = new Rate();
    const interestRate = rateInterface.findRate();

    const repaymentMouthCount = repaymentMouthCountCalc(repaymentPeriod);
	let balanceMinus = balancesMinus(borrowingAmount);
	let balance = borrowingAmount;
	let date = new Date();
    let loanTrans: LoanTran[] = [];
    for (let i:number = 1; i < repaymentMouthCount + 1; i++) {
        const loanTran = buildLoanTran(i, Moment(date).format("2006/01"), fixedRepaymentAmount, interestRate, balance);
		loanTrans = append(loanTrans, loanTran);
		date.setMonth( date.getDate() + 1);
		balance = balanceMinus(loanTran.Principal)
    }

    // 仮でnull返却
    return null
}

export interface RateInterface {
    findRate():{interestRate:number};
}

const buildLoanTran = (id: number, month: string, total: number, 
                       interestRate: number, beforeBalance: number) : LoanTran => {

        const interest = calcMonthInterestCalc(beforeBalance, interestRate);
        const principal = total - interest;
        // 支払後残高が今回支払額より少ない場合は最終支払分と判断しまとめる
        let lastBalance = 0;
        if (beforeBalance-principal-principal < 0) {
            lastBalance = beforeBalance - principal;
        }

        return new LoanTran(id,                                         // id
                            month,                                      // month
                            total + lastBalance,                        // total
                            principal + lastBalance,                    // principal
                            interest,                                   // interest
                            beforeBalance - principal - lastBalance,    // balance
        );
    
}


// 本来ここに記載するものではないが、簡略化のため
const RoundDown = (num: number, place: number) : number => {
	const shift = Math.pow(10, place);
	return Math.floor(num*shift) / shift;
}

// 年利から月ごとの金利を計算
const mouthRateCalc = (interestRate: number): number => {
    return RoundDown(interestRate/100/12, 10);
}

// 年数から支払回数を計算
const repaymentMouthCountCalc = (repaymentPeriod: number) : number => {
	return repaymentPeriod * 12;
}

// 現在の残高と年利から今月の支払い利息を計算
const calcMonthInterestCalc = (balance: number, interestRate: number) : number => {
	return RoundDown(balance*mouthRateCalc(interestRate), 0);
}